source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc

```diff
@@ -123,15 +124,137 @@ AWSLambdaFilter::decodeHeaders(Http::RequestHeaderMap &headers,
+  });
+  response_headers_->setContentLength(buff.length());
+  return Http::FilterDataStatus::Continue;
+}
+
+Http::FilterTrailersStatus 
+AWSLambdaFilter::encodeTrailers(Http::ResponseTrailerMap &) {
+
+  if (functionOnRoute() == nullptr || !functionOnRoute()->unwrapAsAlb()){
+   return Http::FilterTrailersStatus::Continue;
+  }
+  // Future proof against alb http2 support and finalize the data transform
+  const Buffer::Instance&  buff = *encoder_callbacks_->encodingBuffer();
+  encoder_callbacks_->modifyEncodingBuffer([this](Buffer::Instance& enc_buf) {
+    Buffer::OwnedImpl body;
+    parseResponseAsALB(*response_headers_, enc_buf, body);
+    enc_buf.drain(enc_buf.length());
+    enc_buf.move(body);
+  });
+  response_headers_->setContentLength(buff.length());
+  return Http::FilterTrailersStatus::Continue;
+}
+
+
+void AWSLambdaFilter::parseResponseAsALB(Http::ResponseHeaderMap& headers, 
+                    const Buffer::Instance& json_buf, Buffer::Instance& body) {
+
+  ProtobufWkt::Struct alb_response;                
+  try {
+    MessageUtil::loadFromJson(json_buf.toString(), alb_response);
+  } catch (EnvoyException& ex) {
+    ENVOY_LOG(warn, "{}: alb_unwrap set but did not recieve a json payload", 
+                                                   functionOnRoute()->path());
+    headers.setStatus(static_cast<int>(Http::Code::InternalServerError));
+    return;
+  }
+  
+  bool is_bad_format = false;
+  const auto& flds = alb_response.fields();
+  if (flds.contains("body")) {
+    auto rawBody = flds.at("body").string_value();
+    if (flds.contains("isBase64Encoded")){
+      is_bad_format = is_bad_format ||
+                             !flds.at("isBase64Encoded").has_bool_value();
+      if (flds.at("isBase64Encoded").bool_value()){
+        rawBody = Base64::decode(rawBody);
+      }
+    }
+    body.add(rawBody);
+  }
+
+  if (flds.contains("statusCode")){
+    headers.setStatus(flds.at("statusCode").number_value());
+    is_bad_format= is_bad_format || !flds.at("statusCode").has_number_value();
+  }
+  if (flds.contains("headers")){
+    auto hds = flds.at("headers").struct_value();
+    auto hdsFields = hds.fields();
+    for (auto const& hdrEntry : hdsFields) { 
+      headers.addCopy(Http::LowerCaseString(hdrEntry.first),
+                             hdrEntry.second.string_value());
+    }
+  }
+  // While ALB would refuse to parse something with headers + multivalue
+  // Being more permissive in this case was determined to be better.
+  if (flds.contains("multiValueHeaders")){
+     auto hds = flds.at("multiValueHeaders").struct_value();
+    auto hdsFields = hds.fields();
+    for (auto const& hdrEntry : hdsFields) { 
+      auto list = hdrEntry.second.list_value();
+      for (auto const& val : list.values()){
+        headers.addCopy(Http::LowerCaseString(hdrEntry.first),
+                                                            val.string_value());
+      }
+    }
+  }
+  if (!is_bad_format){
```

Comment:
why is the `is_bad_format` done this way? why not just bail early on the first bad format?

Reply:
Mainly done so that we can perform proper error handling in a single place.
Changed the flow as the drain can now be handled in the aggregation of the upper level function.